### ISSUE TITLE
Failed Alignment of sequential stained samples

### ISSUE THREAD

**Issue by rjesud** [NHPatterson/wsireg#50]
I am encountering an issue when aligning 2 images of varying resolution and dynamic range. They are sequentially stained and visually, the structures are extremely similar as expected:
![image](https://user-images.githubusercontent.com/46010245/164358120-4c5d7326-f7ab-49b7-b579-6c222a9cefd6.png)

The left image is IMC and the right is IF. The IMC is encoded as a uint16 images but actually has a pretty small range of intensity values for the nuclear stain (~0-75). The IF is uint16 and has a range of (~100-50000) in the nuclear channel. I use preprocessing 'as_unit8' for both images but contrast enhance for only the IMC. I use 'reg_params=["rigid","affine","nl"]'. Below is the result. 

![image](https://user-images.githubusercontent.com/46010245/164358980-f3edcf8d-2f82-4f40-86e2-d25d664482b4.png)

Do you have recommended preprocessing for this type of alignment?

**Comment by rjesud** [comment]
@NHPatterson Thanks so much for this detailed explanation. It makes a lot of sense and I'm glad I'll be able to look back on this in the documentation (which is great by the way!). 

I did not know the preprocessed images were exported! This will be my first stop in troubleshooting the issue. I have noticed some better results in certain settings when contrast enhancement is turned off. I will also try to limit the transformations to "rigid+nl". 

To answer you questions - This is indeed sequentially stained IF then IMC. So the images represent the exact same tissue. The other big difference I forgot to mention is the fact that the resolutions a quite different. IMC is ~1 um/pixel while IF is ~0.35 um/pixel. I noticed that you have implemented downsampling as a preprocessing step. Would you recommend including this option? For example, downsampling the IF to be in the resolution range of the IMC?

**Comment by NHPatterson** [comment]
@rjesud You shouldn't downsample the images to match resolution. In fact, all image registrations are in _physical_ coordinate space so that multi-scale image registration is possible ([here is a great primer on ITK images, coordinate spaces, image reg.](https://docs.google.com/presentation/d/1Mo_yHmJ7Uai2DKAO2VTyAm56V9KdI-Wx5xVgc9gljmQ/edit#slide=id.p17)). This is why you must provide an image resolution for your image. Make sure you set the `image_res` argument equal to 1 for IMC and 0.35 for IF.  It's extremely important to get these values correct, so double check your image acquisition software if you can't find it in image metadata. Another easy way to find this is to drop the original image into QuPath and check the image information tab. 

The downsampling is more to control memory. If you have extremely 2D large planes that will explode memory as is the case with say a large 40x whole slide brightfield scan), you can downsample by 2x or 4x and `wsireg` will take care to output still at the target resolution. To give you an example.. If you downsampled your IF image by **0.35**, it would have a pixel spacing of **0.7** um / px and then the registration would work by comparing a 1 um/px image to a 0.7 um/px image. But like I said, that doesn't influence the result much, especially because your images are already close in resolution comparatively.

**Comment by rjesud** [comment]
@NHPatterson Thanks for the input! I am still troubleshooting but I think I am narrowing in on the issue. One of the quirks with IMC images is they are written as 32float ometiff. This is done despite the fact that most values are <100. I do a numpy 16 uint conversion but I do not think I am getting integer values. I come back to this conversion. The other quirk is that IMC can have "hot pixels". As an example the vast majority of pixels in a channel can be <20 but there could be 1 or 2 pixels at >10000 (AU). Could it be that this is throwing off the 8bit conversion in wsireg? I do see that some of my preprocessed images are just showing some hot pixels and dropping the lower intensity pixels. Right now, I am exploring CLAHE histogram normalization as a preprocessing step (https://scikit-image.org/docs/stable/auto_examples/color_exposure/plot_equalize.html). I will circle back with progress.

**Comment by NHPatterson** [comment]
> Could it be that this is throwing off the 8bit conversion in wsireg?

Yes this very well could be causing issues. You can turn off that conversion if you see it causing problems (this may be a good example). The `elastix` compiled for python only supports 32-bit float anyway during actual registration, going to uint8 saves memory during all the intensity / spatial preprocessing. Regardless of how you generate the representative image for registration, you can always apply the transforms to the original image and save the new image with the original data type, so any sort of optimization you want for the preprocessing is fine.

**Comment by rjesud** [comment]
Success! It appears the problem was indeed the rescaling of the histogram. I think this can happen at the during contrast enhancement or 8-bit normlization. The odd thing about these IMC images is they are written to 32-bit or 16-bit but the vast majority of the pixel values are <200. There are some pesky high intensity pixels (>10000 AU) that get chosen as the normalization limit which in turn drowns out the relevant signal.  Indeed, when I investigated the preprocessing image I only saw the bright pixels. I tried deactivating contrast enhancement and 8bit normalization but I was still not able to align the images.

So, as a preprocessing step, I eliminate any pixels >99.9 percentile _then_ rescale the histogram. I do this for each channel independently (Does wsireg run the histogram normalization steps on the MIP or each channel before MIP?). After this preprocessing, I can use 8-bit normalization as defined in wsireg. Alignment looks great!

```
from skimage import exposure

img2=(img[i,:,:]<np.percentile(img[i,:,:],99.9))*img[i,:,:]
p1, p99 = np.percentile(img2, (1, 99))
img_rescale = exposure.rescale_intensity(img2, in_range=(p1, p99))
```

Thanks for all the input. This exercise helped me gain a better understanding of the alignment workflow and my image data.

**Comment by NHPatterson** [comment]
That's wonderful to hear! Do you mind sharing an updated screenshot of the overlaid data (It's fun for me to see) and closing the issue when you are satisfied with it?

> Does wsireg run the histogram normalization steps on the MIP or each channel before MIP?

MIP is always first if there is still > 1 channel after data read-in, then everything follows. Your code may have some general usefulness as "hotspot" removal. In the future I may explore adding this as standard or optional preprocessing.
